### DISCUSSION TITLE
Out of boundaries error

### DISCUSSION THREAD

**Discussion by mainarabgouveia** [Parcels-code/Parcels#1086] [Q&A]
Hi,

How do I make my released particles that leave the domain or arrive on the land don't cause my simulations to stop?

I'm getting this error:

INFO: Compiled ArrayJITParticleAdvectionRK45AdvectionDiffusionM1 ==> /tmp/parcels-1000/lib77962480406540cb0477e74d24ef8364_0.so
 98% (14938200.0 of 15120000.0) |####### | Elapsed Time: 0:15:43 ETA:   0:00:11
---------------------------------------------------------------------------
OutOfBoundsError                          Traceback (most recent call last)
/tmp/ipykernel_32755/3033041448.py in <module>
      1 # Now run the remaining ParticleSet (using less accurate Euler Forward for speedup)
      2 
----> 3 pset.execute(pset.Kernel(AdvectionRK45) + AdvectionDiffusionM1,                 # the kernel (which defines how particles move)
      4                       runtime=timedelta(days=175),    # the total length of the run
      5                       dt=-timedelta(minutes=5),      # the timestep of the kernel (negative timestep for backwards run)

~/anaconda3/envs/py3_parcels/lib/python3.9/site-packages/parcels/particleset/baseparticleset.py in execute(self, pyfunc, pyfunc_inter, endtime, runtime, dt, moviedt, recovery, output_file, movie_background_field, verbose_progress, postIterationCallbacks, callbackdt)
    456             # If we don't perform interaction, only execute the normal kernel efficiently.
    457             if self.interaction_kernel is None:
--> 458                 self.kernel.execute(self, endtime=next_time, dt=dt, recovery=recovery, output_file=output_file,
    459                                     execute_once=execute_once)
    460             # Interaction: interleave the interaction and non-interaction kernel for each time step.

~/anaconda3/envs/py3_parcels/lib/python3.9/site-packages/parcels/kernel/kernelsoa.py in execute(self, pset, endtime, dt, recovery, output_file, execute_once)
    225                     recovery_kernel = recovery_map[p.state]
    226                     p.set_state(StateCode.Success)
--> 227                     recovery_kernel(p, self.fieldset, p.time)
    228                     if p.isComputed():
    229                         p.reset_state()

~/anaconda3/envs/py3_parcels/lib/python3.9/site-packages/parcels/tools/statuscodes.py in recovery_kernel_out_of_bounds(particle, fieldset, time)
    194         # TODO: JIT does not yet provide the context that created
    195         # the exception. We need to pass that info back from C.
--> 196         raise OutOfBoundsError(particle, fieldset)
    197     else:
    198         error = particle.exception

OutOfBoundsError: 0
Particle P[4](lon=17.006130, lat=-29.650196, depth=0.000000, time=36208800.000000)
Time: 2021-02-23T02:30:00.000000000,	timestep dt: -3600.000000
Out-of-bounds sampling by particle at (17.006130, -29.650196, 0.000000)

**Comment by erikvansebille** [comment]
Dear @mainarabgouveia. You will have to include a recovery kernel to your execution: 
```python
pset.execute(..., recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle})
``` 

And then create a Recovery Kernel to delete the particle:
```python
def DeleteParticle(particle, fieldset, time):
    particle.delete()
```

See also https://github.com/OceanParcels/parcels/blob/master/parcels/examples/tutorial_Agulhasparticles.ipynb for a full example

**Reply by mollyjames2** [reply]
Hi Erik, 
I'm trying to follow this approach using the following code for pset.execute:
 `pset.execute(pset.Kernel(AdvectionRK4) + DiffusionUniformKh,  runtime=timedelta(days=30),  dt=timedelta(minutes=1),  output_file=output_file, recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle})`

but I'm getting the following error code
``` 
     12 pset.execute(pset.Kernel(AdvectionRK4) + DiffusionUniformKh,                 # the kernel (which defines how particles move)
     13             runtime=timedelta(days=30),    # the total length of the run
     14             dt=timedelta(minutes=1),      # the timestep of the kernel (negative timestep for backwards run)
     15             output_file=output_file,
---> 16             recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle})

NameError: name 'ErrorCode' is not defined
```
the full example you provided gives a 404 not found error. Would be great to get your insights here please. Many thanks

**Reply by erikvansebille** [reply]
Thanks for the question, @mainarabgouveia. But since [Parcels v3.0](https://github.com/OceanParcels/parcels/releases/tag/v3.0.0). we have changed the way that Errors are dealt with. This is not backward-compatible, hence the error when you import `ErrorCode`.

You can now write a Kernel (see example below) and add that as a normal Kernel to your pset.execute():
```python
def DeleteErrorParticle(particle, fieldset, time):
    if particle.state == StatusCode.ErrorOutOfBounds:
        particle.delete()
```
See also [this part of the Kernel Loop tutorial](https://docs.oceanparcels.org/en/latest/examples/tutorial_kernelloop.html#Working-with-Status-Codes) and this comment: https://github.com/OceanParcels/parcels/discussions/1412#discussioncomment-7433532

**Reply by mollyjames2** [reply]
Hi Erik, I have included the DeleteErrorParticle kernel you mentioned above and included it in my set execute 
```` def SampleAge(particle, fieldset, time):
    particle.age += particle.dt 

def Sample_land(particle, fieldset, time):
    particle.on_land = fieldset.landmask[time, particle.depth, particle.lat, particle.lon]

def Particle_death(particle, fieldset, time):
    runtime = 2419200 # 28 days (in seconds)
    if particle.age > runtime:
        particle.delete()
        
def DeleteParticle(particle, fieldset, time):
    particle.delete()

def Distance(particle, fieldset, time):
    lat_dist = (particle.lat - particle.prev_lat) * 1.11e2 # latitudinal distance
    lon_dist = (particle.lon - particle.prev_lon) * 1.11e2 * math.cos(particle.lat * math.pi/180) # longitudinal distance
    particle.distance += math.sqrt(math.pow(lon_dist, 2) + math.pow(lat_dist, 2)) # calculate Euclidean distance
    particle.prev_lon = particle.lon # Set the values for the next time step   
    particle.prev_lat = particle.lat


def DeleteErrorParticle(particle, fieldset, time):
    if particle.state == StatusCode.ErrorOutOfBounds:
        particle.delete()

#Execute an advection kernel on the given fieldset

output_file = pset.ParticleFile(name="/users/modellers/moja/projects/SANH/NEMO-PARCELS/NEMO_parcels_SANH.zarr", outputdt=timedelta(hours=1)) # the file name and the time step of the outputs

# Cutom kernels need to be cast into a kernel object using pset.Kernel before they can function 
kernels = pset.Kernel(AdvectionRK4)+ pset.Kernel(SampleAge)+ pset.Kernel(Sample_land)+pset.Kernel(Distance)+pset.Kernel(DeleteErrorParticle)+pset.Kernel(DiffusionUniformKh)

pset.execute(kernels,              # the kernel (which defines how particles move)
            runtime=timedelta(days=14),    # the total length of the run
            dt=timedelta(minutes=1),      # the timestep of the kernel (negative timestep for backwards run)
            output_file=output_file)
```` 
However, the model is still terminating 
````
---------------------------------------------------------------------------
FieldOutOfBoundError                      Traceback (most recent call last)
Cell In[33], line 15
     12 # Cutom kernels need to be cast into a kernel object using pset.Kernel before they can function 
     13 kernels = pset.Kernel(AdvectionRK4)+ pset.Kernel(SampleAge)+ pset.Kernel(Sample_land)+pset.Kernel(Distance)+pset.Kernel(DeleteErrorParticle)+pset.Kernel(DiffusionUniformKh)
---> 15 pset.execute(kernels,              # the kernel (which defines how particles move)
     16             runtime=timedelta(days=14),    # the total length of the run
     17             dt=timedelta(minutes=1),      # the timestep of the kernel (negative timestep for backwards run)
     18             output_file=output_file)

File ~/miniforge3/envs/parcels/lib/python3.11/site-packages/parcels/particleset.py:948, in ParticleSet.execute(self, pyfunc, pyfunc_inter, endtime, runtime, dt, output_file, verbose_progress, postIterationCallbacks, callbackdt)
    946 # If we don't perform interaction, only execute the normal kernel efficiently.
    947 if self.interaction_kernel is None:
--> 948     res = self.kernel.execute(self, endtime=next_time, dt=dt)
    949     if res == StatusCode.StopAllExecution:
    950         return StatusCode.StopAllExecution

File ~/miniforge3/envs/parcels/lib/python3.11/site-packages/parcels/kernel.py:585, in Kernel.execute(self, pset, endtime, dt)
    583     raise TimeExtrapolationError(p.time)
    584 elif p.state == StatusCode.ErrorOutOfBounds:
--> 585     raise FieldOutOfBoundError(p.lon, p.lat, p.depth)
    586 elif p.state == StatusCode.ErrorThroughSurface:
    587     raise FieldOutOfBoundSurfaceError(p.lon, p.lat, p.depth)

FieldOutOfBoundError: Field sampled out-of-bound, at (81.19326152956675, 5.092092458112055, 0.0)
````

Do you have any insight on why this may be happening?
I am running parcels version 3.0.0

Thanks in advance

**Reply by erikvansebille** [reply]
It's really important that the `DeleteErrorParticle` kernel is the _last_ in you list of kernels. So make it

```python
kernels = [AdvectionRK4, DiffusionUniformKh, SampleAge, Sample_land, Distance, DeleteErrorParticle]
```

See also the last paragraph of https://docs.oceanparcels.org/en/latest/examples/tutorial_kernelloop.html#Working-with-Status-Codes

**Reply by nntaganou** [reply]
Hi Erik,

I am getting an out-of-bound error while using the AdvectionAnalytical advection kernel. Even though I did use the above kernel that deletes the particles at the boundaries, the code still gives me the out-of-bound error for U velocity, while the particle state is still not over 40. 

This is the code:

```
variables = {'U': 'u',
             'V': 'v'}   
dimensions = {'lat': 'Latitude',
              'lon': 'Longitude',
              'time': 'MT'}
 
fieldset = FieldSet.from_xarray_dataset(ds_all_data,variables,dimensions,allow_time_extrapolation=True)
fieldset.mindepth = fieldset.U.depth[0]
fieldset.mindepth = fieldset.U.depth[0]
fieldset.U.interp_method = "cgrid_velocity"
fieldset.V.interp_method = "cgrid_velocity"   

x = lat_value
y = ptcs_lon_all
lons, lats = np.meshgrid(y,x)
psetAA = ParticleSet(fieldset=fieldset, 
                    pclass=ScipyParticle, 
                    lon=lons, 
                    lat=lats,
                    time=0) 

folder_output = '/unity/g2/nntaganou/parcels_gom/'
name = 'test_release_yc_1993_3h_inter_step_daily_data'
file_output = join(f"{folder_output}", f"{name}.zarr")
output = psetAA.ParticleFile(name=file_output, outputdt=timedelta(hours=3))


def CheckOutOfBounds(particle, fieldset, time):
    if particle.state == StatusCode.ErrorOutOfBounds:
        particle.delete()
    
start_date = datetime(1993,1,2,0)
end_date = datetime(1994,1,1,0)
run_time = end_date-start_date
kernels = [AdvectionAnalytical,CheckOutOfBounds]

psetAA.execute(kernels,
            runtime=run_time, 
            dt=timedelta(hours=3), 
            output_file=output)

netcdf_file = join(f'{folder_output}', f'{name}.nc')
ds= xr.open_zarr(f"{file_output}")
```
This is the error:

```
---------------------------------------------------------------------------
FieldOutOfBoundError                      Traceback (most recent call last)
Cell In[13], line 1
----> 1 psetAA.execute(AdvectionAnalytical,
      2             runtime=run_time, 
      3             #endtime=end_date)
      4             dt=timedelta(hours=3), 
      5             output_file=output)

File ~/mambaforge/envs/testenv/lib/python3.11/site-packages/parcels/particleset.py:951, in ParticleSet.execute(self, pyfunc, pyfunc_inter, endtime, runtime, dt, output_file, verbose_progress, postIterationCallbacks, callbackdt)
    949 # If we don't perform interaction, only execute the normal kernel efficiently.
    950 if self.interaction_kernel is None:
--> 951     res = self.kernel.execute(self, endtime=next_time, dt=dt)
    952     if res == StatusCode.StopAllExecution:
    953         return StatusCode.StopAllExecution

File ~/mambaforge/envs/testenv/lib/python3.11/site-packages/parcels/kernel.py:564, in Kernel.execute(self, pset, endtime, dt)
    562     self.execute_jit(pset, endtime, dt)
    563 else:
--> 564     self.execute_python(pset, endtime, dt)
    566 # Remove all particles that signalled deletion
    567 self.remove_deleted(pset)

File ~/mambaforge/envs/testenv/lib/python3.11/site-packages/parcels/kernel.py:543, in Kernel.execute_python(self, pset, endtime, dt)
    540     self.scipy_positionupdate_kernels_added = True
    542 for p in pset:
--> 543     self.evaluate_particle(p, endtime)
    544     if p.state == StatusCode.StopAllExecution:
    545         return StatusCode.StopAllExecution

File ~/mambaforge/envs/testenv/lib/python3.11/site-packages/parcels/kernel.py:628, in Kernel.evaluate_particle(self, p, endtime)
    626 if abs(endtime - p.time_nextloop) < abs(p.dt)-1e-6:
    627     p.dt = abs(endtime - p.time_nextloop) * sign_dt
--> 628 res = self._pyfunc(p, self._fieldset, p.time_nextloop)
    630 if res is None:
    631     if sign_dt*p.time < sign_dt*endtime and p.state == StatusCode.Success:

File <ast>:24, in SetcoordsAdvectionAnalyticalUpdatecoords(particle, fieldset, time)

File ~/mambaforge/envs/testenv/lib/python3.11/site-packages/parcels/field.py:998, in Field.search_indices(self, x, y, z, ti, time, particle, search2D)
    996 def search_indices(self, x, y, z, ti=-1, time=-1, particle=None, search2D=False):
    997     if self.grid.gtype in [GridCode.RectilinearSGrid, GridCode.RectilinearZGrid]:
--> 998         return self.search_indices_rectilinear(x, y, z, ti, time, particle=particle, search2D=search2D)
    999     else:
   1000         return self.search_indices_curvilinear(x, y, z, ti, time, particle=particle, search2D=search2D)

File ~/mambaforge/envs/testenv/lib/python3.11/site-packages/parcels/field.py:822, in Field.search_indices_rectilinear(self, x, y, z, ti, time, particle, search2D)
    820         raise FieldOutOfBoundError(x, y, z, field=self)
    821 if grid.ydim > 1 and (y < grid.lonlat_minmax[2] or y > grid.lonlat_minmax[3]):
--> 822     raise FieldOutOfBoundError(x, y, z, field=self)
    824 if grid.xdim > 1:
    825     if grid.mesh != 'spherical':

FieldOutOfBoundError: U sampled out-of-bound, at (-79.0750963318386, 31.960647583007816, 0.0)
```

The (x, y) at (-79.0750963318386, 31.960647583007816) are the upper bounds of my domain. 

Do you have any ideas as to why I am getting this error? This is the first time I am using analytical advection to advect particles and I am not sure if I am missing something. 

Thank you in advance!

**Reply by erikvansebille** [reply]
I think in this case, the error is in the AnalyticalAdvection scheme. This was coded up as a quick showcase that it would in principle be possible to implement such a scheme within parcels, but wasn't well-tested. I.e. no proper error-handling. You've run into the limits of its capabilities, I fear...

However, my next big project the coming weeks/months is to 'properly' implement AnalyticalAdvection into Parcels, also in JIT-mode. I will then make sure that these out-of-bounds errors are properly treated, too. 

So I suggest you either wait a few more weeks and/or use RungeKutta for the timing being; or write your own removal kernel to remove particles before they even get close to the boundary (if that is even an option for your use-case)

**Reply by nntaganou** [reply]
Thank you for letting me know! 

I might be able to implement a removal kernel after the parcels exit the Straits of Florida -my region is the Gulf of Mexico, so something like that could work.

In your future project, is the fully time-dependent scheme going to be implemented or is it going to be a further development of the stepwise scheme?

Thank you!

**Comment by mainarabgouveia** [comment]
Tks for your help and time.

**Comment by kartadikaria** [comment]
Hi, 

I have a similar problem. I am getting `parcels.tools.statuscodes.FieldOutOfBoundSurfaceError`, when I don't use CheckOutOfBounds/DeleteErrorParticle subroutine. But when I use CheckOutOfBounds/DeleteErrorParticle like the following:

```py
def CheckOutOfBounds(particle, fieldset, time):
    if particle.state == StatusCode.ErrorOutOfBounds:
        particle.delete()

kernels = [parcels.AdvectionRK4,CheckOutOfBounds]

pset.execute(kernels,                 
             runtime=timedelta(hours=72),  
             dt=timedelta(minutes=5),      
             output_file=output_file,
             )
```

I am still getting another error message:

```
  File "/Users/admin/anaconda3/envs/py3_parcels/lib/python3.8/site-packages/parcels/particleset.py", line 849, in execute
    self.kernel = self.Kernel(pyfunc, delete_cfiles=delete_cfiles)
  File "/Users/admin/anaconda3/envs/py3_parcels/lib/python3.8/site-packages/parcels/particleset.py", line 723, in Kernel
    return Kernel.from_list(
  File "/Users/admin/anaconda3/envs/py3_parcels/lib/python3.8/site-packages/parcels/kernel.py", line 483, in from_list
    return functools.reduce(lambda x, y: x + y, pyfunc_list)
  File "/Users/admin/anaconda3/envs/py3_parcels/lib/python3.8/site-packages/parcels/kernel.py", line 483, in <lambda>
    return functools.reduce(lambda x, y: x + y, pyfunc_list)
  File "/Users/admin/anaconda3/envs/py3_parcels/lib/python3.8/site-packages/parcels/kernel.py", line 446, in __add__
    kernel = type(self)(self.fieldset, self.ptype, pyfunc=kernel)
  File "/Users/admin/anaconda3/envs/py3_parcels/lib/python3.8/site-packages/parcels/kernel.py", line 180, in __init__
    self.funccode = funccode or inspect.getsource(pyfunc.__code__)
  File "/Users/admin/anaconda3/envs/py3_parcels/lib/python3.8/inspect.py", line 997, in getsource
    lines, lnum = getsourcelines(object)
  File "/Users/admin/anaconda3/envs/py3_parcels/lib/python3.8/inspect.py", line 979, in getsourcelines
    lines, lnum = findsource(object)
  File "/Users/admin/anaconda3/envs/py3_parcels/lib/python3.8/inspect.py", line 798, in findsource
    raise OSError('could not get source code')
```

If anyone can help?  Thank you.



This is my code:

---------------
```py
import math
from datetime import timedelta as delta
from datetime import datetime,timedelta
from operator import attrgetter
from glob import glob
import matplotlib.pyplot as plt
import numpy as np
import trajan as ta
import xarray as xr
from IPython.display import HTML
from matplotlib.animation import FuncAnimation
import parcels
from parcels import StatusCode


data_folder = '.' 
hydrofiles = sorted(glob(f"{data_folder}/Strait_2004*.nc"))


def DeleteParticle(particle, fieldset, time):
    particle.delete()

def CheckOutOfBounds(particle, fieldset, time):
    if particle.state == StatusCode.ErrorOutOfBounds:
        particle.delete()
        
        
filenames = {
    "U": hydrofiles,
    "V": hydrofiles,
    "W": hydrofiles,        
}
variables = {"U": "UVEL", "V": "VVEL", "W": "WVEL"}

dimensions = {
    "U": {"lon": "XG", "lat": "YG", "depth": "Z", "time": "time"},
    "V": {"lon": "XG", "lat": "YG", "depth": "Z", "time": "time"},
    "W": {"lon": "XG", "lat": "YG", "depth": "Z", "time": "time"},}


fieldset = parcels.FieldSet.from_mitgcm(
    filenames, variables, dimensions, mesh="spherical",
    allow_time_extrapolation=True,)

advection_duration = 48 
ReleaseTime=0
day_start = datetime(2004,1,1,1,0)
day_end = day_start+delta(hours=advection_duration)

nptc=10    
npart=8
lonp = [115.00,115.25,115.5,115.75,116,116.25,116.5,116.75]  
latp = -8.0*np.ones(npart)
lonp=np.array(lonp)
latp=np.array(latp)
depthp = -1.0*np.ones(npart)

releaseTimes = np.arange(0, advection_duration) * delta(hours=1)
releaseTimes += day_start

particles_t     = np.array([])  
particles_lat   = np.array([])
particles_lon   = np.array([])
particles_depth   = np.array([])

for i1,time in enumerate(releaseTimes):
    for i2 in range(nptc):
        particles_t = np.append(particles_t,np.array([(time-day_start).total_seconds() for i in range(len(lonp))]))  
        particles_lat = np.append(particles_lat,latp)
        particles_lon = np.append(particles_lon,lonp)
        particles_depth=np.append(particles_depth,depthp)

        
time_origin = datetime.utcfromtimestamp(np.datetime64(str(fieldset.time_origin)).astype(int)*1e-9)
time_start = day_start - time_origin
particles_t += time_start.total_seconds()


pset = parcels.ParticleSet(fieldset = fieldset,
                      pclass = parcels.JITParticle,
                      lon = particles_lon, 
                      lat = particles_lat,
                      depth=particles_depth,
                      time = particles_t,)

output_file = pset.ParticleFile(name="test.zarr", outputdt=timedelta(minutes=30)) 

kernels = [parcels.AdvectionRK4,CheckOutOfBounds]
pset.execute(kernels,                 
             runtime=timedelta(hours=72),  
             dt=timedelta(minutes=5),      
             output_file=output_file,
             )

```

**Reply by erikvansebille** [reply]
Thanks for reporting @kartadikaria. I'm not sure what's happening here. it seems your `def CheckOutOfBounds(particle, fieldset, time):` is not found properly by the python interpreter. Is all the code above in the same file/script? 

@VeckoTheGecko, do you know what could be going on?

**Reply by kartadikaria** [reply]
Yes Erik, it is the same file.

**Reply by VeckoTheGecko** [reply]
@kartadikaria what sort of environment are you using? I'm interested in the following:
- What version of Parcels (`print(parcels.__version__)`)
- Which OS?
- Are you using Jupyter notebook? A python script?

I notice that you're running on Python 3.8, which means that you likely have an outdated version of Parcels. Newer versions of Parcels (and a lot of the scientific Python ecosystem) have dropped support for Python 3.8 - I recommend that you update to Python 3.10 at the minimum.

**Reply by kartadikaria** [reply]
Hi @VeckoTheGecko, Thanks in advance. I experienced the above problem on 2 PCs.

1. MacOS
- print(parcels.__version__)  : 3.1.2.dev40
- python version: 3.8
- I am using python scripts, not Jupyter
- Just creating a new environment, dedicated to parcels

2. Linux WSL 
- print(parcels.__version__)  : 3.1.2.dev40
- python version: 3.10.12
- I am using python scripts, not Jupyter. I test running line by line.
- mixed environment 

Previously, I used python 3.10 on both machines. However, since I found a similar problem on both of them, then I downgraded the python on one machine. Yet, the problem is still appearing. 

Here the other libraries' version:
>>> print(pd.__version__)
1.4.0
>>> print(np.__version__)
1.26.3
>>> print(xr.__version__)
2023.11.0
>>> print(cartopy.__version__)
0.21.1
>>> print(shapely.__version__)
1.8.5.post1
>>> print(matplotlib.__version__)
3.4.3

---- 

This is the fieldset:

>>> fieldset
<FieldSet>
    fields:
        <Field>
            name            : 'U'
            grid            : RectilinearZGrid(lon=array([ 114.10,  114.11,  114.11, ...,  117.08,  117.09,  117.09], dtype=float32), lat=array([-9.50, -9.49, -9.49, ..., -7.54, -7.54, -7.53], dtype=float32), time=array([ 0.00,  3600.00,  7200.00, ...,  248400.00,  252000.00,  255600.00]), time_origin=2004-01-01T00:30:00.000000000, mesh='spherical')
            extrapolate time: True
            time_periodic   : False
            gridindexingtype: 'mitgcm'
            to_write        : False
        <Field>
            name            : 'V'
            grid            : RectilinearZGrid(lon=array([ 114.10,  114.11,  114.11, ...,  117.08,  117.09,  117.09], dtype=float32), lat=array([-9.50, -9.49, -9.49, ..., -7.54, -7.54, -7.53], dtype=float32), time=array([ 0.00,  3600.00,  7200.00, ...,  248400.00,  252000.00,  255600.00]), time_origin=2004-01-01T00:30:00.000000000, mesh='spherical')
            extrapolate time: True
            time_periodic   : False
            gridindexingtype: 'mitgcm'
            to_write        : False
        <Field>
            name            : 'W'
            grid            : RectilinearZGrid(lon=array([ 114.10,  114.11,  114.11, ...,  117.08,  117.09,  117.09], dtype=float32), lat=array([-9.50, -9.49, -9.49, ..., -7.54, -7.54, -7.53], dtype=float32), time=array([ 0.00,  3600.00,  7200.00, ...,  248400.00,  252000.00,  255600.00]), time_origin=2004-01-01T00:30:00.000000000, mesh='spherical')
            extrapolate time: True
            time_periodic   : False
            gridindexingtype: 'mitgcm'
            to_write        : False
        <VectorField>
            name: 'UV'
            U: <parcels.field.Field object at 0x7f1b75b231c0>
            V: <parcels.field.Field object at 0x7f1b75adfb20>
            W: None
        <VectorField>
            name: 'UVW'
            U: <parcels.field.Field object at 0x7f1b75b231c0>
            V: <parcels.field.Field object at 0x7f1b75adfb20>
            W: <parcels.field.Field object at 0x7f1b75b214b0>

----

I hope, I can get the solution. :)

Thanks.

**Reply by VeckoTheGecko** [reply]
Hi @kartadikaria , 

> * I test running line by line.

Can you elaborate more on what this means? You need to run the script as a file, and not line by line like this. This is a requirement for the `inspect.getsource()` function we use within. For example, the following doesn't work

```
>>> def f(): pass
... 
>>> import inspect
>>> inspect.getsource(f)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/Hodgs004/miniforge3/lib/python3.12/inspect.py", line 1285, in getsource
    lines, lnum = getsourcelines(object)
                  ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Hodgs004/miniforge3/lib/python3.12/inspect.py", line 1267, in getsourcelines
    lines, lnum = findsource(object)
                  ^^^^^^^^^^^^^^^^^^
  File "/Users/Hodgs004/miniforge3/lib/python3.12/inspect.py", line 1096, in findsource
    raise OSError('could not get source code')
OSError: could not get source code
```

Can you try run the Python script? (i.e., `python my_file.py`)

Failing that: it would be good to get this to a minimal example so that we can debug on our end. Can you do the following, seeing that you still get the same errors?:

- Replace the dataset with one from `download_example_dataset`? (like in the tutorial notebooks that we have)
- Delete all code that isn't required to replicate the error
- Can you try using a version of parcels from conda, and see if it replicates? You can create an environment named `new-env` using `conda create -n new-env -c conda-forge python=3.10 parcels`

**Reply by kartadikaria** [reply]
Hi @VeckoTheGecko, I tried to run block-by-block to understand the process. For instance after pset = parcels.ParticleSet() --> output_file ---> kernels ---> pset.execute. The problem occured during pset. execute.

Anyway, after I ran it using "python [my_parcels.py]", as you mentioned.....It had no problem and could finish running. I am still confused, but I accept. Thanks and now I can continue to create other fieldsets in parcels V3.
[Solved]

**Reply by VeckoTheGecko** [reply]
I'm not entirely sure which editor you're using, or how you're running block by block. We recommend that you develop in Jupyter Notebook/Jupyter Lab or by running the script normally via `python ...`
